pkg/sql/schemachanger/scplan/internal/rules/current/dep_statement_phase.go

```diff
@@ -0,0 +1,194 @@
+	clausesForStatementPhaseSingleTransitionForDesc := func(
+		from, to NodeVars,
+		el scpb.Element,
+		targetStatus scpb.TargetStatus,
+		t opgen.Transition,
+		prePrevStatuses []scpb.Status,
+	) rel.Clauses {
+		clauses := rel.Clauses{
+			from.Type(el),
+			to.Type(el),
+			from.El.AttrEqVar(screl.DescID, "_"),
+			from.El.AttrEqVar(rel.Self, to.El),
+			from.Target.AttrEqVar(rel.Self, to.Target),
+			from.Target.AttrEq(screl.TargetStatus, targetStatus.Status()),
+			from.Node.AttrEq(screl.CurrentStatus, t.From()),
+			to.Node.AttrEq(screl.CurrentStatus, t.To()),
+		}
+		// If the stage is revertible and we are in an added state, the statement
+		// phase can have multiple transitions.
+		if t.Revertible() {
+			clauses = append(clauses,
+				from.Target.AttrNeq(screl.CurrentStatus, scpb.Status_DESCRIPTOR_ADDED))
+		}
+		if len(prePrevStatuses) > 0 {
+			clauses = append(clauses,
+				GetNotJoinOnNodeWithStatusIn(prePrevStatuses)(from.Target),
+			)
+		}
+		return clauses
+	}
+	// Ensures that a descriptor dependent is at most allowed a single transition,
+	// if the parent descriptor is in an adding state.
+	clausesForStatementPhaseSingleTransitionForDep := func(
+		from, to NodeVars,
+		el scpb.Element,
+		targetStatus scpb.TargetStatus,
+		t opgen.Transition,
+		prePrevStatuses []scpb.Status,
+	) rel.Clauses {
+		targetDesc := MkNodeVars("target-desc")
+		var descIDVar rel.Var = "desc-id"
+		clauses := rel.Clauses{
+			targetDesc.TypeFilter(rulesVersionKey, isDescriptor),
+			from.Type(el),
+			to.Type(el),
+			targetDesc.JoinTargetNode(),
+			from.DescIDEq(descIDVar),
+			from.El.AttrEqVar(rel.Self, to.El),
+			from.Target.AttrEqVar(rel.Self, to.Target),
+			from.Target.AttrEq(screl.TargetStatus, targetStatus.Status()),
+			from.Node.AttrEq(screl.CurrentStatus, t.From()),
+			to.Node.AttrEq(screl.CurrentStatus, t.To()),
+			targetDesc.DescIDEq(descIDVar),
+			targetDesc.Target.AttrNeq(screl.CurrentStatus, scpb.Status_DESCRIPTOR_ADDED),
+		}
+		if len(prePrevStatuses) > 0 {
+			clauses = append(clauses,
+				GetNotJoinOnNodeWithStatusIn(prePrevStatuses)(from.Target),
+			)
+		}
+
+		return clauses
+	}
+
+	addRules := func(el scpb.Element, targetStatus scpb.TargetStatus) {
+		statusMap := findNoopSourceStatuses(el, targetStatus)
+		if err := opgen.IterateTransitions(el, targetStatus, func(
+			t opgen.Transition,
+		) error {
+			elemName := reflect.TypeOf(el).Elem().Name()
+			ruleName := scgraph.RuleName(fmt.Sprintf(
+				"%s transitions to %s uphold only single transition in statement phase: %s->%s",
+				elemName, targetStatus.Status(), t.From(), t.To(),
+			))
+			registerDepRuleWithMaxPhase(
+				ruleName,
+				scgraph.PreviousStagePrecedence,
+				scop.StatementPhase,
+				"prev", "next",
+				func(from, to NodeVars) rel.Clauses {
+					return clausesForStatementPhaseSingleTransitionForDesc(
+						from, to, el, targetStatus, t, statusMap[t.From()],
+					)
+				},
+			)
+			return nil
```

Comment:
My only real concern with this PR is that there are a LOT of rules added, see how the `deprules` testdata file balloons! 
1. It seems to me that there may be a great many which could be made redundant by re-using the `PreviousTransactionPrecedence` dep-edge kind. Also, if we did that, would we need to introduce this max-phase concept at all?
2. Here do we need to iterate through all transitions? let's stop after one dep-edge is added instead.
3. Independently of the two points above, can we skip elements which never reach their target in more than one transition? I don't think that `IndexColumn` needs a rule when going to public or absent.

Reply:
@postamar Ah good point yeah, this could be a lot less noisy.

1. Reintroducing (1) would allow us to eliminate that concept.
2. Agree we don't need all the permutations think we just need the first step from absent to have this rule
3. Good point